casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -65,31 +66,103 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
     merges(echoContract(1), echoContract(2), Rho("Nil"))
   }
 
-  //same polarity, merges
-  val SAME_POLARITY_MERGE =
-    """
+  def assertSamePolarity(r1: Rho, r2: Rho)(
+      implicit pos: source.Position
+  ) = assert(r1.maybePolarity === r2.maybePolarity, "but same polarity was expected")
+
+  def assertDifferentPolarity(r1: Rho, r2: Rho)(
+      implicit pos: source.Position
+  ) = assert(r1.maybePolarity !== r2.maybePolarity, "but different polarity was expected")
+
+  def assertCrossedPolarity(r1: Rho, base1: Rho, base2: Rho)(
+      implicit pos: source.Position
+  ) =
+    assert(
+      ((r1.maybePolarity === base1.maybePolarity) && (r1.maybePolarity !== base2.maybePolarity)) || ((r1.maybePolarity !== base1.maybePolarity) && (r1.maybePolarity === base2.maybePolarity)),
+      "but crossed polarity was expected"
+    )
+
+  def assertLinear(r1: Rho)(
+      implicit pos: source.Position
+  ) = assert(r1.maybeCardinality === Some(Linear), "expected a linear event")
+
+  def assertOnePersistent(b1: Rho, b2: Rho) =
+    assert(
+      (b1.maybeCardinality === Some(NonLinear)) || (b2.maybeCardinality === Some(NonLinear)),
+      "expected one of the events to be persistent"
+    )
+
+  object SamePolarityMerge {
+    val desc =
+      """
         |Two incoming sends/receives, at most one had a matching dual in TS.
         |The incoming events won't cause more COMMs together (same polarity).
         |They couldn't be competing for the same linear receive/send (at most one had a match).
```

Comment:
consider this:

```
def matches(x: Rho, y: Rho) = {
  val (a, b) = if (x.polarity == Send) (x, y) else (y, x)
  val wildcards = List(F_, C_)
  val knownMatches = 
    List(S0, R0).flatMap(s => (F0 :: C0 :: wildcards).map(r => s -> r)),
    List(S1, R1).flatMap(s => (F1 :: C1 :: wildcards).map(r => s -> r))
  )
 knownMatches.contains(a -> b)
}
```

Now we can incorporate the preconditions based on what matches what